### PR TITLE
I2P support

### PR DESCRIPTION
**Pull request by ignopeverell**
See #2712 for background.

### CODE REVIEW

File: p2p/src/types.rs [hunk 1]
```diff
@@ -106,51 +112,72 @@ impl<T> From<mpsc::TrySendError<T>> for Error {
...
 			let port = reader.read_u16()?;
-			Ok(PeerAddr(SocketAddr::V6(SocketAddrV6::new(
+			Ok(PeerAddr::Socket(SocketAddr::V6(SocketAddrV6::new(
 				Ipv6Addr::new(ip[0], ip[1], ip[2], ip[3], ip[4], ip[5], ip[6], ip[7]),
 				port,
 				0,
 				0,
 			))))
+		} else if addr_format == 2 {
+			let addr = reader.read_bytes_len_prefix()?;
```
Comment (hashmap) [inline p2p/src/types.rs]: `read_bytes_len_prefix` looks dangerous, I'd change it to accept `max_len` as a parameter to be validated before reading the content, can send a tiny pr.

File: p2p/src/types.rs [hunk 2]
```diff
@@ -192,17 +243,56 @@ impl PeerAddr {
...
+		if let PeerAddr::I2p(_) = self {
+			return true;
+		}
+		false
+	}
+
+	/// Whether this is an classic IP address
+	pub fn is_ip(&self) -> bool {
+		if let PeerAddr::I2p(_) = self {
+			return true;
```
Comment (hashmap) [inline p2p/src/types.rs]: shouldn't it be `false`? or just replace the body with `!self.is_i2p()`

File: p2p/src/types.rs (hunk 1 above)
Comment (ignopeverell) [reply p2p/src/types.rs]: Thought the same at first but it actually refuses to read more than 100K. And then that would get you banned right away at the first address read.

File: p2p/src/types.rs (hunk 2 above)
Comment (ignopeverell) [reply p2p/src/types.rs]: Good catch! Will fix.

**Comment by chisa0a** [comment]
What remains on this PR? Would really love to start testing this out, and get Grin running over I2P by default.

Is there anyway I can help?

**Comment by chisa0a** [comment]
@ignopeverell I started work on rebasing this branch on current master (1.1.0-beta2 - b9db129b33), and completing the remaining P2P work (per our discussion on Gitter).

Please let me know if you would like me to submit the rebase work (basically just minor fixups), to make further collaboration a bit easier.

**Comment by chisa0a** [comment]
@ignopeverell ping. I removed the rebase onto 1.1.0 (now it's still on 1.0.3), and completed a large amount of [remaining work for I2P connections](https://github.com/ignopeverell/grin/pull/2).

Please review the work in that PR, and merge if you approve. After the merge, a rebase onto 1.1.0 would be immensely helpful for future collaboration. Thank you for your initial work in supporting I2P.

**Comment by quentinlesceller** [comment]
Closing that PR since we are going in another direction. Hope you'll understand @ignopeverell :)

**Pull request closed without merging**